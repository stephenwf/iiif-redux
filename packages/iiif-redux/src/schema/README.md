# Presentation 2.x and 3.x

This library aims to give full coverage for both presentation 2 and presentation 3 APIs both for importing and selecting content.

When you import Presentation 2 or 3 content it will automatically be detected and imported. Mixed Presentation 2 and 3 content in a single document is
not supported. When importing presentation 2 content, due to inconsistencies you may find the original documents have been modified. This is intentional
to provide a predictable and normalized structure. Descriptive properties also get expanded out to the international specification, defaulting to english,
for these properties. Future exporting tools may revert this during an export. Future exporting will also offer an option to upgrade presentation 2 documents
to the latest stable presentation version.

## API Changes

When presentation updates happen, some parts of the API are deprecated and finally removed, and as such this library does not support calling deprecated
APIs on sources that are not the latest version. For example, between Presentation 2.1 and 3.0 the property `viewingHint` was changed to `behavior`. Along
with this change the scope of the property changed to expand what `behavior` may contain. Because of this, the following behavior applies:

- Calling `getViewingHint` on Presentation 2 document will return `document.viewingHint` (deprecation notice logged in development)
- Calling `getBehavior` on Presentation 2 document will return `document.viewingHint`
- Calling `getViewingHint` on Presentation 3 document will return `null` (deprecation notice logged in development)
- Calling `getBehavior` on Presentation 3 document will return `document.behavior`

Although this library does aim to have full coverage of all presentation versions, it also provides an upgrade path where possible from lower versions of
the specification in your application. If remove all deprecation notices from your application, your application will be compatible with the latest versions
of the specification.

Note: This compatibility only extends to structural differences, and not semantics that may have changed.

## Opting for strict selectors

If you know all of your content conforms strictly to one version of the specification, you can access the selectors directly:

```js
import { canvasByIdSelector } from 'iiif-redux/api/2.x/canvas';
// or
import { canvasByIdSelector } from 'iiif-redux/api/3.x/canvas';

// instead of
import { canvasByIdSelector } from 'iiif-redux/api/canvas';
```

Similarly, if you want to access the IIIF specification compliant selectors (maybe for your own library) you can access them here:

```js
import * as descriptive from 'iiif-redux/api/2.x/iiif/descriptive';
```

Its possible choosing a strict version implementation may improve performance on large documents. This will have undesired effects on the wrong
presentation versions though, so it should only be used on data-sets you control and not driven by user-input like traditional viewers.

## Sequences between 2.x and 3.x

Sequences are a special case in the change between versions 2 and 3. In Presentation 3 you can still call `getSequences` and `getDefaultSequence` on a manifest, but
instead of this being defined in the structure, its a special range that's returned. This acts in the same way as the sequence in presentation 2 and can be used to
reflect the same structure. However, its more common to instead just use the canvases directly off of the manifest. Because of this, presentation 2 has been retro-fitted
with a `getCanvases` selector that will skip the sequence. This is common practice for viewers.

## Language properties

In Presentation 2 and 3 descriptive (human-readable) strings are stored native to their presentation version. When you select, say, a label, you get returned one of 2 structures
representing the string:

```js
// Presentation 2
const p2Label = [{ '@language': 'en', '@value': ['some label'] }];

// Presentation 3
const p3Label = {
  en: ['some label'],
};
```

Its up to the end implementation to support these 2 language implementations. IIIF Redux does offer a function for rendering these strings.

```js
import { intlString } from 'iiif-redux/utility';

const englishString = intlString('en');

// Both return 'some label'
englishString([{ '@language': 'en', '@value': ['some label'] }]);

englishString({
  en: ['some label'],
});
```

## Annotation changes

Annotations have changed a lot between Presentation 2 and 3. It is still unknown how IIIF Redux will differ between implementations. There is a
1-to-1 parity between structural components, which should help to align the changes.

## Paging

Annotation paging defined in W3C remains, but is not supported in Presentation 2 or 3 in this library due to its relatively low usage. This may change
if demand changes.

## Implementation details

When you import a document either manually specifying a presentation version, or if you choose it to be auto-detected, for every resource imported an
index will be created that contains the schema version.

```js
const state = {
  schemaVersions: {
    'https://../1': 3,
    'https://../2': 3,
    'https://../3': 2,
  },
};
```

Although this does add an overhead to the redux store, it ensures that any resource that is loaded is identified correctly when you apply a selector.

You can opt-out of this extra overhead by creating a store targeting a specific version

## Presentation 3 questions

- For the field `seeAlso` and `partOf`, the specification does not provide a list of valid IIIF resources that can be found here, implying all IIIF resources are valid.
- Annotation pages only have one required field `type` which seems counter productive as it cannot function as an annotation page without annotations, should ID also be required?

## Changes to selector framework

Previously if you wanted to get just the label selector, you could do something like this:

```js
import collection from 'iiif-redux/api/collection';

const state = {
  /* ... */
};

const getFirstCollection = () => /***/ null;

const { getLabel } = collection(getFirstCollection);

const label = getLabel(state);

console.log(label);
```

So you could effectively have access to the individual selectors. Because of the uncertainty using multiple presentation versions, this model has now changed to match
the select by ID and collection selectors.

```js
import collection from 'iiif-redux/api/collection';

const state = {
  /* ... */
};

const getFirstCollection = () => /***/ null;

const firstCollection = collection(getFirstCollection);

const selector = firstCollection(api => ({
  label: api.getLabel,
}));

console.log(selector(state).label);

// Alternatively
const getLabel = firstCollection(api => api.getLabel);

const label = getLabel(state);

console.log(label);
```

This unifies how the other selectors work too. If you still want access to individual selectors, say to build up your own library, you can still access each selector
library for each presentation version:

```js
import presentation2Collection from 'iiif-redux/api/2.x/collection';
import presentation3Collection from 'iiif-redux/api/3.x/collection';

const getFirstCollection = () => /***/ null;

const { getLabel, getId } = presentation2Collection(getFirstCollection);

const myCustomP2Selector = createStructuredSelector({
  label: getLabel,
  id: getId,
});
```

This is only recommended if you know what version of the specification you are using as you will have to manage the different versions yourself at this point.

If there is demand for a universal API that exposes single selectors, it may be added and work like this:

```js
import collection from 'iiif-redux/api/2.x-3.x/collection';

const getFirstCollection = () => /***/ null;

const { getLabel, getId } = collection(getFirstCollection);
```

But this is likely to have a negative impact on performance as the version look up will happen for each individual field, instead of once per selector.
