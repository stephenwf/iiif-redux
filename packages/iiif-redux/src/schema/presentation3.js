import { schema, normalize } from 'normalizr';
import { compose } from 'redux';
import addMissingIds from '../compat/addMissingIds';
import parseSelectorTarget from '../utility/parseTargetSelector';

function createEntity(name, processStrategy) {
  const options = {
    idAttribute: 'id',
  };
  if (processStrategy) {
    options.processStrategy = processStrategy;
  }

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections'); // 1
const manifest = createEntity('manifests'); // 2
const canvas = createEntity('canvases'); // 3
const annotation = createEntity('annotations', original => {
  const body = Array.isArray(original.body) ? original.body : [original.body];
  const target = Array.isArray(original.target)
    ? original.target
    : [original.target];
  const newObj = { ...original };
  if (original.body) {
    newObj.body = body;
  }
  if (original.target) {
    newObj.target = target;
  }
  return newObj;
}); // 4
const annotationPage = createEntity('annotationPages'); // 5
const annotationCollection = createEntity('annotationCollections'); // 6
const range = createEntity('ranges'); // 7
const contentResource = createEntity('contentResources'); // 8
const choice = createEntity('choices'); // 9
const canvasReference = createEntity('canvasReferences'); // 10
const selector = new schema.Entity(
  'selectors',
  {
    id: canvas,
  },
  {
    processStrategy(value, parent, key) {
      // @todo This is an impossible path to hit. The process strategy is a strange function that
      // doesn't always hit the containing code. Will need to review at a later date.
      // if (typeof value === 'string') {
      //   return parseSelectorTarget(value);
      // }
      const id = value['@id'] || value.id;
      if (id) {
        return parseSelectorTarget(id);
      }
      return value;
    },
  }
);

// Unofficial types.
const service = createEntity('services');

// Union types
const partOf = new schema.Array(
  {
    collection,
    manifest,
    contentResource,
    annotationCollection,
  },
  entity => {
    switch (entity.type) {
      case 'Collection':
        return 'collection';
      case 'Manifest':
        return 'manifest';
      case 'AnnotationCollection':
        return 'annotationCollection';
      default:
        return 'contentResource';
    }
  }
);

const canvasOrReference = new schema.Union(
  {
    selector,
    canvasReference,
    canvas,
  },
  entity => {
    if (entity && entity.id && entity.id.indexOf('#') !== -1) {
      return 'selector';
    }
    if (entity.type === 'SpecificResource') {
      return 'canvasReference';
    }
    return 'canvas';
  }
);

const annotationBodyMappings = {
  Choice: 'choice',
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const annotationBody = new schema.Union(
  {
    contentResource,
    choice,
  },
  input => {
    if (annotationBodyMappings[input.type]) {
      return annotationBodyMappings[input.type];
    }
    return 'contentResource';
  }
);

const rangeItem = new schema.Union(
  {
    canvas,
    selector,
    range,
    canvasReference,
  },
  input => {
    if (input && input.id && input.id.indexOf('#') !== -1) {
      return 'selector';
    }

    switch (input.type) {
      case 'Canvas':
        return 'canvas';
      case 'Range':
        return 'range';
      case 'SpecificResource':
      default:
        // @todo default?
        return 'canvasReference';
    }
  }
);

// Root resource
/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - Canvas (Recommended)
 * - Annotation (Recommended)
 * - AnnotationPage (Required)
 * - AnnotationCollection (Optional)
 * - Content Resource (Required)
 * - Range (Optional)
 */
const RESOURCE_TYPE_MAP = {
  Collection: 'collection',
  Sequence: 'sequence',
  Manifest: 'manifest',
  Canvas: 'canvas',
  AnnotationCollection: 'annotationCollection',
  AnnotationPage: 'annotationPage',
  Annotation: 'annotation',
  Range: 'range',
  // Content resources.
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const resource = new schema.Union(
  {
    collection,
    manifest,
    canvas,
    annotationCollection,
    annotationPage,
    annotation,
    range,
    contentResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity.type]) {
      return RESOURCE_TYPE_MAP[entity.type];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  // Structural
  items: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity.type === 'Manifest' ? 'manifest' : 'collection')
  ),
  annotations: [annotationPage],

  // Linking.
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra.
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  items: [canvas],
  structures: [range],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,
  start: canvasOrReference,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 3) Canvases
// ===========================================================================
canvas.define({
  // Structural.
  items: [annotationPage],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 4) Annotation
// ===========================================================================
annotation.define({
  // Structural.
  body: [annotationBody],
  target: [selector],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 5) Annotation Collection
// ===========================================================================
annotationCollection.define({
  // Structural.
  items: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // extra
  thumbnail: [contentResource],

  // Paging.
  first: annotationPage,
  last: annotationPage,
});

// ===========================================================================
// 6) Annotation Page
// ===========================================================================
annotationPage.define({
  // Structural.
  items: [annotation],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],

  // Paging.
  next: annotationPage,
  previous: annotationPage,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  // Structural.
  items: [rangeItem],
  annotations: [annotationPage],

  // Linking
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  homepage: contentResource,
  rendering: [contentResource],
  partOf: partOf,
  start: canvasOrReference,
  supplementary: annotationCollection,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 8) Content Resource
// ===========================================================================
contentResource.define({
  // Linking.
  seeAlso: [contentResource],
  service: [service],
  logo: [contentResource],
  rendering: [contentResource],
  partOf: partOf,
  // Structural
  annotations: [annotationPage],
});

// ===========================================================================
// 9) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

// ===========================================================================
// 10) Canvas Reference
// ===========================================================================
canvasReference.define({
  source: canvas,
});

const preprocess = compose(addMissingIds(3.0));

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  collection,
  manifest,
  canvas,
  annotation,
  annotationPage,
  annotationCollection,
  range,
  contentResource,
  choice,
  canvasReference,
  service,
  partOf,
  annotationBody,
  rangeItem,
  RESOURCE_TYPE_MAP,
  resource,
  preprocess,
  normalizeResource as normalize,
};
