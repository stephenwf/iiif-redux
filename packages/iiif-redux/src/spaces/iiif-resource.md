# IIIF Resources
This state is where all of our entities get imported into. The actions here are dynamic
and will call off to other spaces with follow up actions.

## State structure
```js
const state = {
  dereferenced: {
    'http://.../collection.json': { 
      loading: true,
      requested: '00-00-0000:00:00:00', 
      ttl: 600,
      error: null, // or { message: 'Failed to fetch', ... }
    },
  },
  collections: {
    'http://.../collection.json': {/* ... */},
  },
  manifests: {},
  sequences: {},
  canvass: {},
  annotations: {},
  annotationLists: {},
  ranges: {},
  layers: {},
  imageResources: {},
  externalResources: {},
  services: {},
}
```

## Reusable actions
Requesting an item that can definitely be referenced (i.e. manifest):
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    schema: schemas.manifest,
  }
});
```

Requesting an items fields, dereferencing if they don't exist.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    dereferenceIfMissing: ['sequences'],
    schema: schemas.imageService,
  }
});
```

Requesting a resource, with a TTL so if its requested again a fresh copy may be requested.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    schema: schemas.manifest,
  }
});
```

Force-requesting an item, includes cache busting
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    forceFresh: true,
    schema: schemas.manifest,
  }
});
```

## Helper selectors
For use in containers for a more verbose way of showing resources current state while requesting.
```js
isResourceLoaded(id); // returns true/false, for loading screens.
didResourceError(id); // returns error object or null
```

## Potential sagas
In each of the spaces for the different resource types, we can have different reactions and
store different state. 

```js
function* manifestSaga() {
  yield takeEvery('MANIFEST_SUCCESS', handleSuccess);
} 

function* handleSuccess({ payload: { id } }) {
  // Set the current manifest to the one we just requested.
  yield put(setCurrentManifest(id));
  
  // Grab the state + start canvas
  const state = yield select();
  const startCanvas = getStartCanvas(state);

  // Set the canvas if its set.
  if (startCanvas) {
    yield put(setCurrentCanvasIndex(startCanvas)); // Another action creator.
  }
}
```
