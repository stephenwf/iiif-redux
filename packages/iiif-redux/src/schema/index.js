import * as presentation2 from './presentation2';
import * as presentation3 from './presentation3';

export { presentation2, presentation3 };

// Default is 2 until 3 is ratified?
export default presentation2;
