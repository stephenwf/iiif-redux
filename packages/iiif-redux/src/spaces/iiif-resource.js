import cachedFetch from 'fetch-unless-cached';
import { createActions, handleActions } from 'redux-actions';
import validUrl from 'valid-url';
import { call, put, select, takeEvery, all } from 'redux-saga/effects';
import deepmerge from 'deepmerge';
import update from 'immutability-helper';
import * as presentation2 from '../schema/presentation2';
import * as presentation3 from '../schema/presentation3';
import trackPresentationVersion from '../utility/trackPresentationVersion';
import {
  COLLECTION_ERROR,
  COLLECTION_REQUEST,
  COLLECTION_SUCCESS,
} from './collections';
import {
  MANIFEST_ERROR,
  MANIFEST_REQUEST,
  MANIFEST_SUCCESS,
} from './manifests';
import getPresentationVersionFromResource from '../utility/getPresentationVersionFromResource';
const debug = require('debug')('iiif-redux');

const IIIF_RESOURCE_REQUEST = 'IIIF_RESOURCE_REQUEST';
const IIIF_RESOURCE_SUCCESS = 'IIIF_RESOURCE_SUCCESS';
const IIIF_RESOURCE_ERROR = 'IIIF_RESOURCE_ERROR';
const IIIF_RESOURCE_REQUEST_UNKNOWN = 'IIIF_RESOURCE_REQUEST_UNKNOWN';
const IIIF_RESOURCE_IDENTIFY = 'IIIF_RESOURCE_IDENTIFY';

const DEFAULT_MAPPINGS = {
  collection: [COLLECTION_REQUEST, COLLECTION_SUCCESS, COLLECTION_ERROR],
  manifest: [MANIFEST_REQUEST, MANIFEST_SUCCESS, MANIFEST_ERROR],
};

const {
  iiifResourceRequestUnknown,
  iiifResourceRequest,
  iiifResourceSuccess,
  iiifResourceError,
} = createActions({
  [IIIF_RESOURCE_REQUEST_UNKNOWN]: (
    resourceId,
    options = {},
    mappings = DEFAULT_MAPPINGS
  ) => ({
    resourceId,
    dereferenceIfMissing: options.dereferenceIfMissing || false,
    ttl: options.ttl || 600,
    forceFresh: options.forceFresh,
    mappings,
  }),
  [IIIF_RESOURCE_REQUEST]: (resourceId, types, schema, options = {}) => ({
    resourceId,
    types,
    schema,
    dereferenceIfMissing: options.dereferenceIfMissing || false,
    ttl: options.ttl || 600,
    forceFresh: options.forceFresh,
  }),
  [IIIF_RESOURCE_SUCCESS]: (resourceId, normalizedResponse) => ({
    resourceId,
    normalizedResponse,
  }),
  [IIIF_RESOURCE_ERROR]: (resourceId, error) => ({
    resourceId,
    error,
  }),
});

const DEFAULT_STATE = {
  dereferenced: {},
  resources: {
    collections: {},
    sequences: {},
    manifests: {},
    canvases: {},
    canvasReferences: {},
    annotations: {},
    annotationLists: {},
    annotationPages: {},
    annotationCollections: {},
    ranges: {},
    layers: {},
    imageResources: {},
    externalResources: {},
    contentResources: {},
    choices: {},
  },
};

const resourceReducer = handleActions(
  {
    [iiifResourceSuccess]: (state, { payload: { normalizedResponse } }) =>
      deepmerge(state, normalizedResponse),
  },
  DEFAULT_STATE.resources
);

const dereferencedReducer = handleActions(
  {
    [iiifResourceRequest]: (state, { payload: { resourceId, ttl } }) =>
      update(state, {
        [resourceId]: {
          $set: {
            resourceId,
            ttl,
            requested: state[resourceId]
              ? state[resourceId].requested
              : new Date(),
            loading: state[resourceId] ? state[resourceId].loading : true,
          },
        },
      }),
    [iiifResourceRequestUnknown]: (state, { payload: { resourceId, ttl } }) =>
      update(state, {
        [resourceId]: {
          $set: {
            resourceId,
            ttl,
            loading: state[resourceId] ? state[resourceId].loading : true,
          },
        },
      }),

    [iiifResourceError]: (state, { payload: { resourceId, error } }) =>
      update(state, {
        [resourceId]: {
          loading: { $set: false },
          error: { $set: error },
        },
      }),

    [iiifResourceSuccess]: (state, { payload: { resourceId } }) =>
      update(state, {
        [resourceId]: {
          loading: { $set: false },
        },
      }),
  },
  DEFAULT_STATE.dereferenced
);

async function requestResource(resourceId, options) {
  /* istanbul ignore if */
  if (process.env.NODE_ENV === 'production') {
    return cachedFetch(resourceId);
  }

  return fetch(resourceId).then(resp => resp.json());
}

function* errorAction(type, resourceId, error) {
  yield put(iiifResourceError(resourceId, error));
  if (type) {
    yield put({ type, payload: { resourceId, error } });
  }
}

function* successAction(type, resourceId, normalizedResponse) {
  yield put(iiifResourceSuccess(resourceId, normalizedResponse));
  yield put({ type, payload: { resourceId, normalizedResponse } });
}

function* requestIiifResource({ payload }) {
  const { resourceId, types, ...options } = payload;
  if (types.length !== 3 && process.env.NODE_ENV !== 'production') {
    yield call(
      errorAction,
      null,
      resourceId,
      'DEV ERROR: You must pass in exactly 3 action types'
    );
  }

  const [REQUEST, SUCCESS, ERROR] = types;

  if (!validUrl.isWebUri(resourceId)) {
    yield call(
      errorAction,
      ERROR,
      resourceId,
      'Resource is not a valid Web URI.'
    );
    return;
  }

  const state = yield select();

  // @todo logic for cache.
  if (
    state.dereferenced[resourceId] &&
    state.dereferenced[resourceId].loading ===
      false /*&& options.forceFresh === false*/
  ) {
    debug('Skipping fetch for resource %s using cache.', resourceId);

    return;
  }

  yield put({ type: REQUEST, payload });

  try {
    debug('Fetching resource %s', resourceId);
    const response = yield call(requestResource, resourceId, options);

    const { result, entities } = yield call(
      importResource,
      response,
      resourceId
    );
    debug('Finished normalize resource %s', result.id);

    yield call(successAction, SUCCESS, result.id, entities);
  } catch (err) {
    debug('Error: %O', err);
    yield call(errorAction, ERROR, resourceId, err);
  }
}

function* importResource(response, resourceId) {
  const version = getPresentationVersionFromResource(response);
  if (version === 2) {
    return yield call(importPresentation2Resource, response, resourceId);
  }
  if (version === 3) {
    return yield call(importPresentation3Resource, response, resourceId);
  }

  throw new Error('Unknown IIIF Presentation version');
}

function* importPresentation2Resource(response, resourceId) {
  if (response['@id'] !== resourceId) {
    // @todo this will not catch "partOf" fields.
    debug(
      'Resource ID does not match requested presentation 2 resource, patching... Found: %s Expected: %s',
      response['@id'],
      resourceId
    );
    response['@id'] = resourceId;
  }

  debug('Starting normalize presentation 2 resource %s', resourceId);

  return trackPresentationVersion(
    2,
    presentation2.normalize(response, presentation2.resource)
  );
}

function* importPresentation3Resource(response, resourceId) {
  if (response.id !== resourceId) {
    // @todo this will not catch "partOf" fields.
    debug(
      'Resource ID does not match requested presentation 3 resource, patching... Found: %s Expected: %s',
      response.id,
      resourceId
    );
    response.id = resourceId;
  }

  debug('Starting normalize presentation 3 resource %s', resourceId);

  return trackPresentationVersion(
    3,
    presentation3.normalize(response, presentation3.resource)
  );
}

function* getResourceType(resourceId, response) {
  const presentationVersion = getPresentationVersionFromResource(response);
  if (presentationVersion === 2) {
    return yield call(getPresentation2ResourceType, resourceId, response);
  }
  if (presentationVersion === 3) {
    return yield call(getPresentation3ResourceType, resourceId, response);
  }

  throw new Error('Unknown IIIF Presentation version');
}

function* getPresentation2ResourceType(resourceId, response) {
  if (!presentation2.RESOURCE_TYPE_MAP[response['@type']]) {
    throw new Error('Unknown Presentation 2 resource type');
  }

  return presentation2.RESOURCE_TYPE_MAP[response['@type']];
}

function* getPresentation3ResourceType(resourceId, response) {
  if (!presentation3.RESOURCE_TYPE_MAP[response.type]) {
    throw new Error('Unknown Presentation 3 resource type');
  }

  return presentation3.RESOURCE_TYPE_MAP[response.type];
}

function* requestUnknownResource({
  payload: { resourceId, mappings, ...options },
}) {
  if (!validUrl.isWebUri(resourceId)) {
    yield put(iiifResourceError(resourceId, 'Resource is not a valid URL.'));
    return;
  }

  try {
    // 1) Request the resource.
    debug('Fetching unknown resource %s', resourceId);
    const response = yield call(requestResource, resourceId, options);
    if (!response) {
      yield put(iiifResourceError(resourceId, 'Unknown resource'));
      return;
    }

    const type = yield call(getResourceType, resourceId, response);

    if (!mappings[type]) {
      yield put(
        iiifResourceError(
          resourceId,
          `Resource type is not in configured mappings (${Object.keys(
            mappings
          ).join(', ')})`
        )
      );
      return;
    }

    // 3) Grab constants from a map (to be created).
    const [REQUEST, SUCCESS, ERROR] = mappings[type];

    // 4) Send to regular import (will be cached).
    yield put(iiifResourceRequest(resourceId, [REQUEST, SUCCESS, ERROR]));
  } catch (err) {
    yield put(iiifResourceError(resourceId, err.message));
  }
}

function* saga() {
  yield all([
    takeEvery(IIIF_RESOURCE_REQUEST, requestIiifResource),
    takeEvery(IIIF_RESOURCE_REQUEST_UNKNOWN, requestUnknownResource),
  ]);
}

export {
  saga,
  resourceReducer,
  dereferencedReducer,
  iiifResourceRequest,
  iiifResourceRequestUnknown,
  DEFAULT_STATE,
  IIIF_RESOURCE_REQUEST,
  IIIF_RESOURCE_SUCCESS,
  IIIF_RESOURCE_ERROR,
  IIIF_RESOURCE_REQUEST_UNKNOWN,
  IIIF_RESOURCE_IDENTIFY,
};
