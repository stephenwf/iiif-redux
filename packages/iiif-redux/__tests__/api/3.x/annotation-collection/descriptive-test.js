import annotationCollection from '../../../../src/api/3.x/annotation-collection';
import { normalize } from '../../../../src/schema/presentation3';
import fixtures from './fixtures';
import { DEFAULT_STATE } from '../../../../src/spaces/iiif-resource';

describe('api/3.x/annotation-collection/technical', () => {
  fixtures.forEach(({ fixture, name }) => {
    describe(`AnnotationCollection: ${name}`, () => {
      const {
        result: { id },
        entities,
      } = normalize(fixture);
      const state = {
        resources: { ...DEFAULT_STATE.resources, ...entities },
      };
      const selectAnnotationCollection = s =>
        s.resources.annotationCollections[id];

      const {
        getLabel,
        getMetadata,
        getSummary,
        getThumbnailId,
        getThumbnail,
        getRequiredStatement,
        getRights,
      } = annotationCollection(selectAnnotationCollection);

      test('getLabel', () => {
        expect(getLabel(state)).toMatchSnapshot(`getLabel ${name}`);
      });

      test('getMetadata', () => {
        expect(getMetadata(state)).toMatchSnapshot(`getMetadata ${name}`);
      });

      test('getSummary', () => {
        expect(getSummary(state)).toMatchSnapshot(`getSummary ${name}`);
      });

      test('getThumbnailId', () => {
        expect(getThumbnailId(state)).toMatchSnapshot(`getThumbnailId ${name}`);
      });

      test('getThumbnail', () => {
        expect(getThumbnail(state)).toMatchSnapshot(`getThumbnail ${name}`);
      });

      test('getRequiredStatement', () => {
        expect(getRequiredStatement(state)).toMatchSnapshot(
          `getRequiredStatement ${name}`
        );
      });

      test('getRights', () => {
        expect(getRights(state)).toMatchSnapshot(`getRights ${name}`);
      });
    });
  });
});
