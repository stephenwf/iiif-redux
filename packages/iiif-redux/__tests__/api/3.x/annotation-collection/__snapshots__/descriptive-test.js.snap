// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Full Annotation Collection getLabel: getLabel Full Annotation Collection 1`] = `
Object {
  "en": Array [
    "Annotation Collection 1",
  ],
}
`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Full Annotation Collection getMetadata: getMetadata Full Annotation Collection 1`] = `
Array [
  Object {
    "label": Object {
      "en": Array [
        "Author",
      ],
    },
    "value": Object {
      "@none": Array [
        "Anne Author",
      ],
    },
  },
  Object {
    "label": Object {
      "en": Array [
        "Published",
      ],
    },
    "value": Object {
      "en": Array [
        "Paris, circa 1400",
      ],
      "fr": Array [
        "Paris, environ 1400",
      ],
    },
  },
  Object {
    "label": Object {
      "en": Array [
        "Notes",
      ],
    },
    "value": Object {
      "en": Array [
        "Text of note 1",
        "Text of note 2",
      ],
    },
  },
  Object {
    "label": Object {
      "en": Array [
        "Source",
      ],
    },
    "value": Object {
      "@none": Array [
        "<span>From: <a href=\\"https://example.org/db/1.html\\">Some Collection</a></span>",
      ],
    },
  },
]
`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Full Annotation Collection getRequiredStatement: getRequiredStatement Full Annotation Collection 1`] = `
Object {
  "label": Object {
    "en": Array [
      "Attribution",
    ],
  },
  "value": Object {
    "en": Array [
      "Provided by Example Organization",
    ],
  },
}
`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Full Annotation Collection getRights: getRights Full Annotation Collection 1`] = `"https://creativecommons.org/licenses/by/4.0/"`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Full Annotation Collection getSummary: getSummary Full Annotation Collection 1`] = `
Object {
  "en": Array [
    "Book 1, written be Anne Author, published in Paris around 1400.",
  ],
}
`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Full Annotation Collection getThumbnail: getThumbnail Full Annotation Collection 1`] = `
Object {
  "id": "https://example.org/images/book1-page1/full/80,100/0/default.jpg",
  "service": Array [
    "https://example.org/images/book1-page1",
  ],
  "type": "Image",
}
`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Full Annotation Collection getThumbnailId: getThumbnailId Full Annotation Collection 1`] = `"https://example.org/images/book1-page1/full/80,100/0/default.jpg"`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Minimal Annotation Collection getLabel: getLabel Minimal Annotation Collection 1`] = `null`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Minimal Annotation Collection getMetadata: getMetadata Minimal Annotation Collection 1`] = `Array []`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Minimal Annotation Collection getRequiredStatement: getRequiredStatement Minimal Annotation Collection 1`] = `null`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Minimal Annotation Collection getRights: getRights Minimal Annotation Collection 1`] = `null`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Minimal Annotation Collection getSummary: getSummary Minimal Annotation Collection 1`] = `null`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Minimal Annotation Collection getThumbnail: getThumbnail Minimal Annotation Collection 1`] = `null`;

exports[`api/3.x/annotation-collection/descriptive AnnotationCollection: Minimal Annotation Collection getThumbnailId: getThumbnailId Minimal Annotation Collection 1`] = `null`;
